### PR TITLE
Density matrix backend

### LINKED ISSUE #31: Density matrix backend
**Is your feature request related to a problem? Please describe.**
We cannot simulate with noise/error at the moment.

**Describe the solution you'd like**
Add density matrix backend so that we can simulate various quantum channels (ones described by Kraus ops).

**Describe alternatives you've considered**
na

**Additional context**
Consider how this could be done in TN backend.

### PR DESCRIPTION
**Pull request by shinich1**
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `pytest`)
- format added code and tests by `black -l 120 <filename>`
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
 
Then, please fill in below:

**Context (if applicable):**
na

**Description of the change:**
Add density matrix simulator backend and interface for custom noise model.

**Related issue:**
closes #31

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

### CODE REVIEW

File: tests/test_density_matrix.py
Comment (shinich1) [inline tests/test_density_matrix.py:379]: what is this testing?

File: tests/test_density_matrix.py
Comment (shinich1) [inline tests/test_density_matrix.py:396]: could you add tests for larger simulations (see tests for pattern and sv sim)? compare results with circuit.simulat_statevector at no-noise case, and with noisy case, fix the np.random.seed and check the results with some known result you calculate without using the backend?

File: tests/test_density_matrix.py
Comment (shinich1) [inline tests/test_density_matrix.py:308]: how about np.ones((4, 4))/2. ?

File: tests/test_density_matrix.py
Comment (nabe98) [reply tests/test_density_matrix.py:379]: This is checking the concrete case where both resulting states becomes equal(no dephase case(p=0) and p=1 case).

File: tests/test_density_matrix.py
Comment (nabe98) [reply tests/test_density_matrix.py:308]: If you mean np.ones((2,2))/2, can be refactored as one-liner. Thanks! (in the next commit I'll push this change)
`expected_matrix_1 = np.kron(np.array([[1, 0], [0, 0]]), np.ones((2, 2)) / 2)`

w.r.t this, `expected_matrix_2` is also refactored as one-liner
`expected_matrix_2 = np.kron(np.array([[0, 0], [0, 1]]), np.array([[0.5, -0.5], [-0.5, 0.5]]))`

File: tests/test_density_matrix.py
Comment (nabe98) [reply tests/test_density_matrix.py:432]: I added the test case for 3-qubit QFT. Please see that if it's enough.

Comment (shinich1) [reply tests/test_density_matrix.py:432]: could you use https://github.com/TeamGraphix/graphix/blob/master/tests/random_circuit.py#L56 instead?

Comment (nabe98) [reply tests/test_density_matrix.py:432]: After the commit [a6745bb](https://github.com/TeamGraphix/graphix/pull/64/commits/a6745bb6e09a039bab0d8b30d88f7d79294d7165), I'll fix this part.

Comment (nabe98) [reply tests/test_density_matrix.py:432]: What do you mean by this? I think it's not the test for `dephase` if you mean something like `np.testing.assert_allclose(dm_backend.state.fidelity(random_state), dm_backend.state.fidelity(exact_qft_state))` instead of
`np.testing.assert_allclose(dm_backend.state.fidelity(noisy_state), dm_backend.state.fidelity(exact_qft_state))`

Comment (shinich1) [reply tests/test_density_matrix.py:432]: I mean something like `np.testing.assert_allclose(1, dm_backend.state.fidelity(known_noisy_state_for_noise_realization))` (fix the noise realizatino by np.random.seed())

**Review by thierry-martinez — COMMENTED** [review]
I didn't have time to reread all the code yet, so here are only very minor comments. I hope I will have time to read more code in the following days!

File: .gitignore
Comment (shinich1) [inline .gitignore:15]: do we need to specify them?

File: graphix/channels.py
Comment (shinich1) [inline graphix/channels.py:172]: how about adding two qubit dephasing channel and pauli channel?

Comment (mgarnier59) [reply graphix/channels.py:172]: Random Pauli channel added in [5501114](https://github.com/TeamGraphix/graphix/pull/64/commits/5501114f4c9135bb49fee29db37721836a9e71be). 
Isn't the 2-qubit dephasing channel just the tensor product of single-qubt dephasing channels? Indedd, some channel manipulation utilities could be added (tensor, composition).

File: .gitignore
Comment (mgarnier59) [reply .gitignore:15]: Just my local files that I wanted to ignore.

File: graphix/channels.py
Comment (shinich1) [reply graphix/channels.py:172]: great, how does it work? shall we move it to `channels` instead, that seems useful?

Comment (mgarnier59) [reply graphix/channels.py:172]: How does what work?

I have no opinion on whether to move the random Pauli channels to `channel` or leave it in `random_objects` to me it made sense to gather all random stuff at the same place. 

I planned to add all Channel manipulation stuff a bit later.

Comment (mgarnier59) [reply graphix/channels.py:172]: > > How does what work?
> 
> `rand_Pauli_channel_kraus` function.
You select the number of operators you want in the Channel (rank) from all possible Pauli tensor products that are precomputed, you build a discrete probability distribution with this number of possibilities and then you choose randomly the correct number of operators.


> Let's move things in `random_objects.py` outside of `tests`, perhaps? functions in `random_circuit` is not so generalizable (beyond unit tests) but I feel that things in `random_objects.py` that you added could be quite useful.

Agreed. Random channels can also be useful.

File: graphix/channels.py
Comment (nabe98) [inline graphix/channels.py:70]: This is better since, if `python assert check_rank(kraus_data)` fails, the first two lines are meaningless.

```python
        assert check_rank(kraus_data)
        self.nqubit = int(np.log2(kraus_data[0]["operator"].shape[0]))
        self.kraus_ops = kraus_data
```

File: setup.py
Comment (nabe98) [inline setup.py:40]: Please use " instead of ' if there is no reason!

**Review by nabe98 — COMMENTED** [review]
Hello! I've left some simple comments (@shinich1 asked me to). Please check them.
As soon as I have time, I might leave some more comments.

File: graphix/channels.py
Comment (mgarnier59) [reply graphix/channels.py:70]: Ok. We'll also change all dimensions check for something more efficient soon.

File: setup.py
Comment (mgarnier59) [reply setup.py]: Don't know what you mean. If I remember correctly, random is no longer needed. I'll remove it.

File: graphix/random_objects.py
Comment (masa10-f) [inline graphix/random_objects.py:22]: How about using `scipy.stats.unitary_group`?
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.unitary_group.html

Comment (mgarnier59) [reply graphix/random_objects.py:22]: Used it in the beginning then tried to remove it to avoid using `scipy` but couldn't and used this more simple version. I opted for the easy version since I had no specific requirement on the measure used for the sampling but we might want to make sure that we use the Haar measure.

File: graphix/random_objects.py
Comment (mgarnier59) [inline graphix/random_objects.py:1]: Moved this file to `graphix/` as suggested by @shinich1 . This actually implies that `scipy`has to be added to the requirements. @shinich1 wasn't a huge fan of adding requirements, so what do we decide?

File: graphix/channels.py
Comment (nabe98) [reply graphix/channels.py:16]: I think it is appropriate to classify our implementation as trace-preserving.
In my understanding, a scenario becomes non-trace-preserving (NTP) when projection measurements are conducted, although there might exist other NTP cases. In our implementation, measurement operations are not represented by Kraus operators; instead, they are handled using a different method.

Comment (mgarnier59) [reply graphix/channels.py:16]: CPTP maps represent what happens when  a **unitary** evolution is applied on the larger system system tensor ancilla (or environment) and you trace out the ancilla part. If measurements are introduced, then you recover the theory of POVM which still must sum to 1 (completeness). As far as I can tell, we don't need non-trace preserving maps.

File: graphix/sim/density_matrix.py
Comment (shinich1) [reply graphix/sim/density_matrix.py:239]: I agree with Maxine, let us leave it like this for now.

File: graphix/random_objects.py
Comment (shinich1) [reply graphix/random_objects.py:1]: I missed that point that I asked! 
I think, at this stage, having random_objects toolkit far exceeds the dependency being a bit heavier. shall we add `scipy` then?

File: graphix/channels.py
Comment (mgarnier59) [reply graphix/channels.py:16]: I agree, let's postpone this discussion if we realize we need to implement non TP maps (with I don't believe is necessary so far). The photon-loss channel is a CPT map but this requires modelling that is not currently done at the graphix level.

File: graphix/random_objects.py
Comment (mgarnier59) [reply graphix/random_objects.py:1]: I also think so. Random objects might be interesting for benchmarking algorithms outside of tests.

File: graphix/channels.py
Comment (masa10-f) [reply graphix/channels.py:16]: @mgarnier59 @shinich1 @nabe98 Thanks! I agree to address it in another issue

File: graphix/random_objects.py
Comment (masa10-f) [reply graphix/random_objects.py:1]: I'd like to note that `quimb` , tensor network library, already depends on `scipy`, so there will be no actual dependency change :)

Comment (mgarnier59) [reply graphix/random_objects.py:1]: Great!

File: graphix/random_objects.py
Comment (masa10-f) [reply graphix/random_objects.py:22]: I've read that Haar-random Unitary can be created using QR decomposition. Original paper is [1] and it is also cited in `scipy.stats.unitary_group` module. So, I believe it would be better to use `unitary_group` module!

[1] Francesco Mezzadri [How to generate random matrices from the classical compact groups](https://arxiv.org/abs/math-ph/0609050)
[2] [Understanding the Haar measure](https://pennylane.ai/qml/demos/tutorial_haar_measure/#mezzadri2006)

Comment (mgarnier59) [reply graphix/random_objects.py:22]: As you want. I'm not sure it matters that much anyways.

**Review by shinich1 — COMMENTED** [review]
All the code look good! Docs rendering and example(sphinx gallery) fixed.

Once test passes, we can squash and merge to master. I can't seem to approve  because I created this PR. Can you approve this PR now, @masa10-f ?

once approved, could you squash and merge, @mgarnier59 , so that you'll become the committer?

**Review by masa10-f — APPROVED** [review]

**Pull request merged**
